### ISSUE TITLE
Generic thumbnailer

### ISSUE THREAD

**Issue by nodiscc** [shaarli/Shaarli#111]
Currently we only support thumbnails for a bunch of sites. Thumbnailers for other sites have been requested at https://github.com/sebsauvage/Shaarli/issues/91, https://github.com/sebsauvage/Shaarli/issues/127, https://github.com/sebsauvage/Shaarli/issues/169, #14. As we cannot possibly add regexes for every site out there, I think a generic thumbnailer is the way to go. IMHO it should

 * Run server-side to allow caching and not burden client bandwidth and CPU
 * Work on every page that shows images via HTML/CSS.
 * Have little impact on performance (little CPU/bandwidth overhead for the server, don't slow down page load)
 * Still allow for regexes for specific sites to speed up the thumbnailing process
 * Be portable (no exotic dependencies like WSGI scripts, node...)
 * Be optional (can be disabled through a checkbox)

@pikzen has a nice prototype of such a thumbnailer in php at https://github.com/pikzen/Shaarli/blob/thumbnailer/ThumbnailExtractor.php (184 lines). You can try it by cloning it and pointing your browser to `http://path/to/your/ThumbnailExtractor.php?surface=no&url=http://example.url/with/images`.

It works by downloading the page HTML, parsing it to find `<img>` tags, and assigning a score to each one depending on regexes on it's URL. We are free to use the highest-scored image as a thumbnail. It can then be cached by Shaarli. The `?surface={yes,no}` param sets whether the server should download images from the page and calculate their surface, which affects their score. It has a performance impact.

I'd like to see this as a plugin, but there is no way to store the thumbnail URL in the link array from a plugin, so we would have to call `tpl/plugins/thumbnailer/thumbnailer.php?url=http://blahblah` 50 times per pageload, and it would not be cacheable.

Alternatively we could have an option for this thumbnailer to _only_ get favicons and not try to get thumbnails from the whole page. It has been requested too.

Please comment on how/if you'd like to see this implemented. The code is about the same complexity/size as the previous thumbnailer code, and works with all sites that contain images.

**Comment by nodiscc** [comment]
My bad https://github.com/sebsauvage/Shaarli/issues/91 is about reducing size of thumbnails and should be addressed separately.

**Comment by pikzen** [comment]
A few things about this thumbnailer:

 * Can't be done clientside (in Javascript) because of CORS rules. A project exists, ImageResolver which does it in JS but it basically sends a request to self, which sends a request to a specialized server to handle thumbnailing, which returns the data. It's a bunch of complications, and in the end it still runs on the server.
 * Doing this with surface checking is AWFULLY slow. I'm talking 30s slow on certain websites.
 * Without surface checking, it is still pretty damn slow at 2.5s per request. Or 0.00045 if using specialized codepaths like YouTube's (no need to do a query, just generate the URL). While I'm not a big fan of this, adding specialized codepaths for common websites would make it faster in most cases. I'd hardly call 2.5s to add a link "acceptable".
 * ThumbnailExtractor sends back an array of images on the site, with a score from 0 to 1, 1 being what seems the most important image. One way of speeding this up would be to stop querying once we've reached a score lower than 0.5, and adding maluses the further down the page the image is.

Basically, I'm focused on making it fast right now, but I fear that as soon as we have to make requests, it's going to be 1s+.

You can also test it at http://ec2-54-93-76-97.eu-central-1.compute.amazonaws.com/Shaarli/ThumbnailExtractor.php?url=http://reddit.com/

Note that it totally breaks down when http:// is not present and that warnings may be shown.

**Comment by ArthurHoaro** [comment]
There is a warning if the url doesn't contain any image.

About performance issue, I'm not sure that's a huge problem since the requests will be done once, when the link is added. After that, we just need cache.

Another solution would be to make the requests asynchronous. It can't be done in JS, but we can call the thumbnailer through AJAX.

**Comment by pikzen** [comment]
Yeah,  the parameters parts is just here for demo purposes and does no checks or anything.

I'd love to push out an improved version (which doubles performance, down to 1.5s per request WITH surface checking.) but it's currently sitting on a computer with no SSH or physical access :(

**Comment by nodiscc** [comment]
@pikzen I'd remove the youtube-specific code and use the `<meta property="og:image" content="https://i.ytimg.com/vi/cqCoUvz_nwI/maxresdefault.jpg">` to get thumbnails for youtube. #117 also proposed using `og:image`. This metatag is described on the [Opengraph website](http://ogp.me/). This may also be faster than parsing the whole, page; if the metatag points to a valid image, use this and exit.

**Comment by mro** [comment]
You also may want to have a look at 
- thumbnail caching in https://github.com/mro/Shaarli/commit/92cddfd40b72ea97aa042dea487d6e2162fc4b4a and 
- cleaning up broken images https://github.com/mro/Shaarli/commit/c48c07ba6d07e1162d09c94cad256a5c543b1d55

**Comment by nodiscc** [comment]
@mro nice to have you here.

> generic vs. site specific and eager GET vs. lazy

I'd rather have it use eager `GET` requests instead of adding long site-specific rules (as you said in https://github.com/shaarli/Shaarli/pull/117#issuecomment-75344614, most of your code is trying to resolve image URLs, which will become a mess if we add more and more sites - removing all site-specific code, [including current one](https://github.com/shaarli/Shaarli/blob/master/index.php#L1980) would be good). Yes `og:image` is a gift. We may need to set a timeout for the `GET` requests to complete.

> thumbnail caching

and cache expiration look very good (expiration time may be discussed)

> cleaning up broken images

I'd enjoy this. `unlink($filepath);` makes me cautious though, and we'd better try all possible cases to make sure it doesn't `unlink(something_unexpected)`.

> make various relative URLs absolute

Yes I've hit bugs related to relative URLs with @pikzen's thumbnailer in a few cases. This should be sorted out.

> implement following http redirects

In what cases is this useful? Can you provide an URL? Presumably if you bookmark a page your shaared link should point to the redirection endpoint. I'm not sure we should follow redirects server-side. Rogue redirects/hijacking? Cross-domain redirects?

Thanks again for your input

**Comment by mro** [comment]
> @mro nice to have you here.

that's very kind, thank you, it's very nice to be welcome.

>> implement following http redirects

> In what cases is this useful? Can you provide an URL?

When bookmarking URLs that promise long-term availability. They're often logical URLs redirecting to a 'real' URL. See http://www.w3.org/Provider/Style/URI for the idea of permanent URLs.

example: http://purl.org/dc/terms/ (*purl* is for *permanent url*)

As to eager `GET` to look for `og:image`: Maybe the 'add link' form might get 1 more field - a (optional) thumbnail image url.  
If that resolves there's no `GET` to the URL to bookmark needed - only `GET` the image for resize+caching. Otherwise it may be fine to just probe for a few like `<meta`, `<link rel` or just any any (1st) image URL inside the html.

**Comment by pikzen** [comment]
>just any any (1st) image URL inside the html.
This usually gives awful results. On most of the websites I tried, it returned icons, or in the best case the website's logo. Even google.com's first image is the 'Download Chrome' logo.

I 100% agree on one point: parsing every single image should be a last resort because it's awfully long. A way to make this invisible for the end user would be to run in on another thread (lol php threads), but in most cases og:image <link>s provide all we need.
I'll be pushing the updated code in a few hours once I get access to it.

file_get_contents does follow the redirect on http://purl.org/dc/terms/

**Comment by e2jk** [comment]
Important to remember is that we are not a thumbnailing system, but a links
system. It's nice to have a thumbnail when relevant, or when the link is to
an actual image. But we shouldn't try to absolutely get an image for each
link (another reason not to parse the entire HTML for images).

I suggest we go with og:image as a generic way, and have a couple of
well-known websites handled differently in case they don't provide that tag
(e.g. Youtube, Vimeo, etc. - not that I've checked if they provide og, but
as an example of the types of websites we should look at to provide
thumbnails)

**Comment by pikzen** [comment]
I'm actually looking at releasing my thumbnailer as a small library once I've ironed out a few edge cases. So Shaarli can make use of a thumbnailing system for its links system :3

>(another reason not to parse the entire HTML for images).
After profiling I saw that the actual request (as in the part in file_get_content that actually downloads the page) is 99.02% of all the thumbnailing time, while parsing it is barely anything. And considering that we need to download them anyways to parse the head, it might as well do a small looking on the images. Surface checking is disabled by default to prevent huge load times.

Major websites like Youtube ,Vimeo and others do provide og:image just because they are often shared on Facebook and others. The websites that don't are the small websites, blogs and others which already don't have a special case (and thus don't have any image displayed on Shaarli).

**Comment by mro** [comment]
>>>> implement following http redirects

>>> Can you provide an URL?

>> example: http://purl.org/dc/terms/ (*purl* is for *permanent url*)

> https://github.com/shaarli/Shaarli/issues/111#issuecomment-75500548: `file_get_contents` does follow the redirect on http://purl.org/dc/terms/

Ouch, indeed. But it fails for http://heise.de/-2557279 or http://spiegel.de/article.do?id=1019990.

**Comment by pikzen** [comment]
Are you sure you don't have an issue with your PHP config (or mine has a generous config) ? I can file_get_contents both of those URLs and they return the entire site without any issue.

**Comment by nodiscc** [comment]
Same results as @pikzen on Debian stable, apache 2.2.22-13+deb7u4, libapache2-mod-php5 5.4.36-0+deb7u3, mostly vanilla config.

**Comment by mro** [comment]
You both (https://github.com/shaarli/Shaarli/issues/111#issuecomment-75734661, https://github.com/shaarli/Shaarli/issues/111#issuecomment-75728058) are absolutely right - what fooled me was that the `$httpstatus` is `301 Moved Permanently`. So no match here: https://github.com/mro/Shaarli/blob/92cddfd40b72ea97aa042dea487d6e2162fc4b4a/index.php#L2583.

Thank you for your patience!

**Comment by nodiscc** [comment]
@mro I see you're still working on your thumbnailing system. To prevent duplicate work, can you either:

1. Base your work on @pikzen's `thumbnailer` branch
2. Open a separate issue or pull request detailing your thumbnailer implementation.

Regarding client-side caching: what do you mean? Of course your browser caches thumbnails as any other resource. The caching should be done server-side to prevent redoing thumbnailing operations again and again.

Regarding resizing: I see we're not recommending `php-gd` in the README yet. Without GD resizing won't work.

 - [ ] TODO: add `php-gd` as an optional dependency in the README.

**Comment by mro** [comment]
@ https://github.com/shaarli/Shaarli/issues/111#issuecomment-76140684
> To prevent duplicate work

I don't mind duplicate work. Increases choice.

> Regarding client-side caching: what do you mean? Of course your browser caches thumbnails as any other resource.

No AFAIK. It's actively disabled: https://github.com/shaarli/Shaarli/blob/dbcad7406eedaeba259a6e1584ba3b0823115c8d/index.php#L86 vs. https://github.com/mro/Shaarli/commit/b30a147d3a4ac80571f761b00adb5f963044bb9f

**Comment by mro** [comment]
IMO the core (unanswered) question is: **what is the default for links - thumbnail or not?**

Currently it's *no thumb* - only after a positive string match the whole process starts. This saves space and clutter (few thumbnails = few server-side cached files) but requires domain/url whitelists or patterns. Or an additional field (thumb source url or at least a flag) in the data store.

Secondary question is whether to create & cache a thumbnail image or steal bandwidth. Again - either whitelists or persistent flag per link.

I hesitated to anticipate any decision and focused on code reduction & clarifying of the existing code base without introducing any new constructs or fiddling with persistence for now.

What do you guys think?

**Comment by nodiscc** [comment]
@mro Your pull request at https://github.com/shaarli/Shaarli/issues/128 looks good.

> I don't mind (a bit of) duplicate work. Increases choice.

Yes fine! As long as you keep a pull request up to keep track of your advancement.

 * nice `$filepath` and `$blankpath` clarification/cleanup.
 * `function saveThumbnailToCacheAndRespond()` very much needed.
 * generic thumbnailing function using `meta` looks good (can you detail/explain what the regexes are supposed to return?)

Question: Can we get rid of site-specific thumbnailers?

 * Eg. the youtube thumbnailer can be removed, as your generic thumbnailing code will properly pick up the `og:image` property on youtube.
 * Can we get rid of all other site-specific code? (`youtube.com, youtu.be, pix.toile-libre.org, imgur.com, i.imgur.com, dailymotion.com, .imageshack.us, flickr.com, vimeo.com, ted.com, xkcd.com`) And use parsing of `<img src=` tags to [assign points to each image](https://github.com/pikzen/Shaarli/blob/thumbnailer/ThumbnailExtractor.php#L173), like https://github.com/pikzen/Shaarli/blob/thumbnailer/ThumbnailExtractor.php does? (eg a `.*\/comics\/.*\.png` match would get +2 points) _Note that this would only apply to sites where there are no Opengraph meta tags._

**Comment by mro** [comment]
> can you detail/explain what the regexes are supposed to return?

[they](https://github.com/mro/Shaarli/blob/d1fa472fbbcbab4d3881dc8c1c8f04476e90de32/index.php#L2443) return the thumbnail image url candidate. It's 2 matching groups because the attributes can be written `<meta property=... content=.../>` or `<meta content=... property=.../>` (`<link` likewise) and exactly one matching group is non-null in case of a match.

A proper xml (sax/dom) parser would be cleaner, but also increase conceptual complexity. A few plain regexp matches look acceptable to me. Doing more than simple probing for the usual suspects is too much magic for my taste.
As somebody mentioned earlier, shaarli isn't about thumbnails in the first place. Rather about links and simplicity. So having simpler code and the same thumbnails as yesterday is IMO ok, more thumbnails are bonus.
Frankly I'm not a fan of https://github.com/pikzen/Shaarli/blob/thumbnailer/ThumbnailExtractor.php#L173

More important is IMO to explicitly answer https://github.com/shaarli/Shaarli/issues/111#issuecomment-76151435.

**Comment by nodiscc** [comment]
> More important is IMO to explicitly answer #111 (comment).

Default should be enabled, with a toggle-off option. I was initially in favor of a plugin for this (see https://github.com/shaarli/Shaarli/pull/129 for a basic plugin system) but in the current form it does not allow us to cache files (does it?).

> more thumbnails are bonus

Yes your code improves the current thumbnailer, even though I'd still like to remove the ~200 lines of site-specific code, it's better than what we have. We should consider merging (after testing). Thanks for your work.

**Comment by nodiscc** [comment]
New thumbnailer Pull Request in https://github.com/shaarli/Shaarli/pull/153, based on previous #128 which broke because of a missing character. It's now unbroken, but thumbnailing code itself doesn't work. Help welcome

**Comment by mro** [comment]
the thumbnailer does work - you got bitten by https://github.com/shaarli/Shaarli/issues/111#issuecomment-76151435. Default as of today: no thumbnail! That's what you see.

**Comment by nodiscc** [comment]
Pull Request at #153 is working. Please help rewiewing/testing/adding comments.

**Comment by nodiscc** [comment]
Should we move this to the 0.9beta milestone? The generic thumbnailer is working, with a few exceptions (see https://github.com/shaarli/Shaarli/pull/153). We could start by whitelisting only a bunch of known working sites and merging this for 0.9beta; then fixing the regex and allowing users to select `Generate thumbnails for all domains` in the prefs for 1.1.

Ok?

**Comment by mro** [comment]
> a few exceptions (see #153)

what the remaining issues?

**Comment by nodiscc** [comment]
Regexes failing loudly for some links (https://github.com/shaarli/Shaarli/pull/153#issuecomment-78573468), `alt` text still displayed (https://imgur.com/YtiJhbW)

**Comment by mro** [comment]
you applied https://github.com/shaarli/Shaarli/pull/153#issuecomment-78706017 and still get the same errors?

**Comment by nodiscc** [comment]
No sorry, I didn't read your message properly. I'll do what you suggest and try again.
